kernel/src/expressions/column_names.rs

```diff
@@ -106,6 +129,131 @@ impl Hash for ColumnName {
     }
 }
 
+/// A wrapper for `ColumnName` that pretty-prints the column name with fields delimited by periods
+/// and enclosed in square brackets, e.g. `column_name!("a.b.c")` would print as `"[a].[b].[c]"`.
+// TODO: Marked as test-only for now, as a demonstration. Should this be Display for ColumnName, so
+// `ColumnName::new(["a", "b.c", "d"])` would print as `"a.[b.c].d"` instead of merely `"a.b.c.d"`?
+#[cfg(test)]
+struct FancyColumnName(ColumnName);
+
+#[cfg(test)]
+impl Display for FancyColumnName {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let mut delim = None;
+        for s in self.0.iter() {
+            use std::fmt::Write as _;
+
+            // Only emit the period delimiter after the first iteration
+            if let Some(d) = delim {
+                f.write_char(d)?;
+            } else {
+                delim = Some('.');
+            }
+
+            if s.is_empty() || s.contains(|c: char| !c.is_ascii_alphanumeric() && c != '_') {
+                // Special situation detected. For safety, surround the field name with brackets
+                // (with proper escaping if the field name itself contains brackets).
+                f.write_char('[')?;
+                for c in s.chars() {
+                    match c {
+                        ']' => f.write_str("]]")?,
```

Comment:
I think we should write `[` as `[[` as well, even though technically it's unambiguous, reading a string like:

`a.[b[c]].d].e` is _hard_, and `a.[b[[c]].d].e` feels easier.

It does complicate parsing a bit though

Reply:
I do worry about parsing complexity, especially when I suspect that square brackets are vanishingly rare. Is it really worth the trouble?

Reply:
Also, if we do something it has to be unambiguous. Are we _requiring_ that `[` _always_ be doubled up inside field names, and a single `[` throws an error about trying to start a new field name without closing the previous one?

Reply:
Yeah, I would imagine that any field name with a single `[` in it would be an error. If you wanted that you'd need to do `[field[[name]`.

I'm okay with having it this way though since, as you note, this is probably a rare occurence.

Reply:
Update: The parser and to-string changes were actually quite simple, so I went ahead and did it.